File: go/ironfleet/tests/services/security_service_test.go
package services

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"ironfleet/services/security"
	"testing"
)

func TestValidateCommandAuthCorrectSignature(t *testing.T) {
	secret := "test-secret-key"
	command := "DEPLOY:fleet-alpha"
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(command))
	sig := hex.EncodeToString(mac.Sum(nil))
	if !security.ValidateCommandAuth(command, sig, secret) {
		t.Fatal("expected valid auth")
	}
}

func TestCheckPathTraversalBlocksDotDot(t *testing.T) {
	if !security.CheckPathTraversal("/safe/path/file.txt") {
		t.Fatal("expected safe path to pass")
	}
	if security.CheckPathTraversal("../../etc/passwd") {
		t.Fatal("expected path traversal to be blocked")
	}
}

func TestRateLimitCheckRespectsLimit(t *testing.T) {
	if !security.RateLimitCheck(5, 10, 60) {
		t.Fatal("expected under limit")
	}
	if security.RateLimitCheck(10, 10, 60) {
		t.Fatal("expected at limit to be blocked")
	}
}

func TestComputeRiskScoreInRange(t *testing.T) {
	score := security.ComputeRiskScore(3, true, true)
	if score < 0 || score > 1.0 {
		t.Fatalf("score out of range: %f", score)
	}
}

File: js/nebulachain/tests/services/security.service.test.js
const test = require('node:test');
const assert = require('node:assert/strict');
const crypto = require('node:crypto');
const security = require('../../services/security/service');

test('validateCommandAuth verifies HMAC signature', () => {
  const secret = 'test-secret-key-1234';
  const command = 'deploy:v2';
  const sig = crypto.createHmac('sha256', secret).update(command).digest('hex');
  const result = security.validateCommandAuth({ command, signature: sig, secret, requiredRole: 'admin', userRoles: ['admin'] });
  assert.equal(result.authorized, true);
});

test('checkPathTraversal detects dot-dot attacks', () => {
  assert.equal(security.checkPathTraversal('../../etc/passwd').safe, false);
  assert.equal(security.checkPathTraversal('/safe/path').safe, true);
});

test('rateLimitCheck enforces request limits', () => {
  const result = security.rateLimitCheck({ requestCount: 200, limit: 100, windowS: 60 });
  assert.equal(result.limited, true);
});

test('computeRiskScore factors failed attempts', () => {
  const score = security.computeRiskScore({ failedAttempts: 5, geoAnomaly: true, timeAnomaly: false });
  assert.ok(score > 50);
});
